File: 2022/day20.py
with open("day20.in") as f:
    lines = list(f.readlines())


def solve(part):
    nums = [(i, int(x) * [1, 811589153][part - 1]) for (i, x) in enumerate(lines)]
    m = len(nums) - 1
    for _ in range([1, 10][part - 1]):
        for I in range(len(nums)):
            i = [x[0] for x in nums].index(I)
            n = nums[i]
            j = (n[1] + i) % m
            del nums[i]
            nums.insert(j, n)

    n0 = [x[1] for x in nums].index(0)

    print(sum(nums[(n0 + d) % len(nums)][1] for d in [1000, 2000, 3000]))


solve(1)
solve(2)

File: 2022/day20.ts
import { getInput, clog, int, sum, range } from "./utils";

let dataString = getInput("day20.in");

function mod(x: number, m: number) {
  return ((x % m) + m) % m;
}

function solve(part: 1 | 2) {
  const nums = dataString
    .split(/\n/g)
    .map((x) => [int(x) * (part === 1 ? 1 : 811589153)]) as [number][];
  let startNums = [...nums];
  range(part === 1 ? 1 : 10).forEach((_) => {
    for (let n of startNums) {
      const i = nums.indexOf(n);
      const j = mod(n[0] + i, nums.length - 1);
      nums.splice(i, 1);
      nums.splice(j, 0, n);
    }
  });
  const v = nums.map((x) => x[0]);

  const n0 = v.indexOf(0);

  clog(sum([1000, 2000, 3000].map((d) => v[mod(n0 + d, v.length)])));
}

solve(1);
solve(2);
